### PR TITLE
feat(hybrid-cloud): Add user and team fields to grouphistory and exter…

### PR DESCRIPTION
**Pull request by vbro**
The is the first pass at removing user actors and making `sentry_actor` a region silo model, and hence removing the concept of actors from the control silo.

This PR adds a `user` field to `sentry_grouphistory` and `user` and `team` fields to `sentry_externalactor`, and writes to those fields.

Next steps:
* Add `user` and `team` fields to `sentry_notificationsetting`. This may involve splitting the `sentry_notificationsetting` model into two, one for the region silo and the other for the control silo.
* Make a data migraiton to backfill the new `user` and `team` fields.
* Read from `user` or `team` fields and remove references to actor from the three models.
* Remove the signal that creates sentry_actor records for newly created users.

### CODE REVIEW

File: src/sentry/api/bases/external_actor.py [hunk 1]
```diff
@@ -127,6 +152,13 @@ def validate_team_id(self, team_id: int) -> Team:
         except Team.DoesNotExist:
             raise serializers.ValidationError("This team does not exist.")
 
+    def to_internal_value(self, data: Mapping[str, Any]) -> OrderedDict[str, Any]:
+        internal_value: OrderedDict[str, Any] = super().to_internal_value(data)
+        team_id = internal_value["team_id"]
+        if isinstance(team_id, Team):
+            internal_value["team"] = team_id
```
Comment (markstory) [inline src/sentry/api/bases/external_actor.py:159]: Why is `team_id` an instance of `Team`? Shouldn't it be an int?

File: src/sentry/migrations/0349_actor_user_direct_reference.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+            field=sentry.db.models.fields.foreignkey.FlexibleForeignKey(
+                null=True, on_delete=django.db.models.deletion.CASCADE, to=settings.AUTH_USER_MODEL
+            ),
+        ),
+        migrations.AddField(
+            model_name="grouphistory",
+            name="user",
+            field=sentry.db.models.fields.foreignkey.FlexibleForeignKey(
+                null=True, on_delete=django.db.models.deletion.SET_NULL, to=settings.AUTH_USER_MODEL
+            ),
```
Comment (markstory) [inline src/sentry/migrations/0349_actor_user_direct_reference.py:48]: Should we be adding another foreign key here? This looks like a key that would need to be broken in the future as it is between user (control silo) and group history (region silo).

File: src/sentry/migrations/0349_actor_user_direct_reference.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+                null=True, on_delete=django.db.models.deletion.CASCADE, to="sentry.Team"
+            ),
+        ),
+        migrations.AddField(
+            model_name="externalactor",
+            name="user",
+            field=sentry.db.models.fields.foreignkey.FlexibleForeignKey(
+                null=True, on_delete=django.db.models.deletion.CASCADE, to=settings.AUTH_USER_MODEL
+            ),
+        ),
```
Comment (markstory) [inline src/sentry/migrations/0349_actor_user_direct_reference.py:42]: Should we be adding another foreign key here? This looks like a key that would need to be broken in the future as it is between user (control silo) and group history (region silo).

Comment (vbro) [reply src/sentry/migrations/0349_actor_user_direct_reference.py:42]: Yes, `GroupHistory.user` and `ExternalActor.user` foreign keys will have to be broken later, and I have included those in the exemptions in `test_silo_models.py`.

I think it's simpler to deal with breaking the foreign keys to the User model this way than to do that indirectly via the Actor model.

Part of this work is also to make the Actor model a region-only silo model (as opposed to having different region and control silo versions) or remove it altogether.

File: src/sentry/migrations/0349_actor_user_direct_reference.py (hunk 2 above)
Comment (vbro) [reply src/sentry/migrations/0349_actor_user_direct_reference.py:48]: See comment above.

File: src/sentry/api/bases/external_actor.py (hunk 1 above)
Comment (vbro) [reply src/sentry/api/bases/external_actor.py:159]: This seems to be the default behavior with `rest_framework.serializers.ModelSerializer` which says that 
> The process of automatically determining a set of serializer fields based on the model fields is reasonably complex ...

See: https://github.com/encode/django-rest-framework/blob/master/rest_framework/serializers.py#L864-L879

Perhaps the default behavior can be changed with `ExternalTeamSerializer.Meta.fields`, but those fields are still necessary before backfilling all records with the team/user.

**Review by wedamija — COMMENTED** [review]
Could you add some context here on why the actor model needs to be removed? It simplifies a lot of constraints to have just the one foreign key.

**Comment by vbro** [comment]
> Could you add some context here on why the actor model needs to be removed? It simplifies a lot of constraints to have just the one foreign key.

TLDR: The Actor abstraction adds complexity related to hybrid-cloud and in most references the actor can only be a user or team and not both.

For Hybrid Cloud, all models are in either the region or control silo and not both.
Actor points to either a Team (region silo) or a User (control silo). So if we keep the Actor model in the control silo, then region silo models that reference Actor will have to first make a cross-silo query to figure out what type the actor is. Hence, for instance, a cross-silo query just to reference a Team object which is in the same silo. And vice-versa if we keep the Actor model in the region silo.

The other option is to have two different versions of the Actor model in region and control silos. This involves syncing the actor table data across silos, guessing that if an actor is not available in the current silo then it could be in another silo, or creating actor ids that are not synced across silos.

The other point with the Actor model is that in 3 of the 5 [models that reference it](https://sentry-schema.netlify.app/tables/sentry_actor.html#Relationships), the actor can only be a user or team, not both. (Rule and Alert rule expects the actor to be a Team, and GroupHistory expects the actor to be a User).
For the other two models (ExternalActor and NotificationSetting) the actor can be either a User or Team. Currently, both GroupAssignee and GroupOwner have both user and team fields instead of actor. And it is possible to use a check constraint to ensure that one and only one of the two fields is not null.

There is [more discussion on the topic here](https://www.notion.so/sentry/Actor-bottleneck-f304d764041a45c4933e98df044506de). Comments and suggestions are welcome.

**Comment by wedamija** [comment]
> > Could you add some context here on why the actor model needs to be removed? It simplifies a lot of constraints to have just the one foreign key.
> 
> TLDR: The Actor abstraction adds complexity related to hybrid-cloud and in most references the actor can only be a user or team and not both.
> 
> For Hybrid Cloud, all models are in either the region or control silo and not both. Actor points to either a Team (region silo) or a User (control silo). So if we keep the Actor model in the control silo, then region silo models that reference Actor will have to first make a cross-silo query to figure out what type the actor is. Hence, for instance, a cross-silo query just to reference a Team object which is in the same silo. And vice-versa if we keep the Actor model in the region silo.
> 
> The other option is to have two different versions of the Actor model in region and control silos. This involves syncing the actor table data across silos, guessing that if an actor is not available in the current silo then it could be in another silo, or creating actor ids that are not synced across silos.
> 
> The other point with the Actor model is that in 3 of the 5 [models that reference it](https://sentry-schema.netlify.app/tables/sentry_actor.html#Relationships), the actor can only be a user or team, not both. (Rule and Alert rule expects the actor to be a Team, and GroupHistory expects the actor to be a User). For the other two models (ExternalActor and NotificationSetting) the actor can be either a User or Team. Currently, both GroupAssignee and GroupOwner have both user and team fields instead of actor. And it is possible to use a check constraint to ensure that one and only one of the two fields is not null.
> 
> There is [more discussion on the topic here](https://www.notion.so/sentry/Actor-bottleneck-f304d764041a45c4933e98df044506de). Comments and suggestions are welcome.

Fair enough, I think I hoped we'd take the time to push actor into more tables and get rid of the dual fks, but if it's causing problems then this seems fine. Is the idea to remove it from all tables, or just tables in control?

File: src/sentry/migrations/0349_actor_user_direct_reference.py (hunk 3 above)
Comment (markstory) [reply src/sentry/migrations/0349_actor_user_direct_reference.py:42]: We could add the columns without the DB constraints now though. That would save us having to break them in the future. The drawback of no constraints is that we won't benefit from referential integrity and won't be able to leverage the SET_NULL/CASCADE behaviors. However we won't have those properties long term so we might as well adapt our application logic now.
